### PR TITLE
Eth 535 pull earnings into operator contract

### PR DESCRIPTION
**Pull request by samigabor**
ticket: https://linear.app/streamr/issue/ETH-535/poolvalue-update-mechanism-should-pull-earnings-into-operatorcontract

### CODE REVIEW

File: packages/network-contracts/contracts/OperatorTokenomics/Operator.sol [hunk 1]
```diff
@@ -754,31 +752,32 @@ contract Operator is Initializable, ERC2771ContextUpgradeable, IERC677Receiver,
...
-        emit PoolValueUpdate(totalValueInSponsorshipsWei, token.balanceOf(address(this)));
+        
+        _redelegateOperatorsShare(sumEarnings);
 
-        // if total difference is more than allowed, then slash the operator a bit: move some of their pool tokens to reward the caller
-        // TODO: this could move pool tokens to someone who isn't delegated into the pool! TODO: Add them if they're not in the pool?
+        // if earnings are more than allowed, then slash the operator a bit: move some of their pool tokens to reward the caller
         uint allowedDifference = getApproximatePoolValue() * streamrConfig.poolValueDriftLimitFraction() / 1 ether;
-        if (sumActual > sumApprox + allowedDifference) {
+        if (sumEarnings > allowedDifference) {
```
Comment (samt1803) [inline packages/network-contracts/contracts/OperatorTokenomics/Operator.sol:771]: deduct brokers share here, also add a test, that test edge case values

**Review by samt1803 — CHANGES_REQUESTED** [review]
also change punish test to use operators share

File: packages/network-contracts/contracts/OperatorTokenomics/Operator.sol (hunk 1 above)
Comment (samigabor) [reply packages/network-contracts/contracts/OperatorTokenomics/Operator.sol]: Updated

File: packages/network-contracts/contracts/OperatorTokenomics/Operator.sol [hunk 2]
```diff
@@ -326,24 +324,23 @@ contract Operator is Initializable, ERC2771ContextUpgradeable, IERC677Receiver,
         }
         uint cashoutWei = sponsorship.reduceStakeTo(targetStakeWei);
         stakedInto[sponsorship] -= cashoutWei;
-        updateApproximatePoolvalueOfSponsorship(sponsorship);
         emit StakeUpdate(sponsorship, sponsorship.stakedWei(address(this)));
+        totalValueInSponsorshipsWei -= cashoutWei;
+        emit PoolValueUpdate(totalValueInSponsorshipsWei, token.balanceOf(address(this)));
```
Comment (jtakalai) [inline packages/network-contracts/contracts/OperatorTokenomics/Operator.sol:329]: event was missing? Good catch 👍

**Review by jtakalai — APPROVED** [review]
👍

File: packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts [hunk 3]
```diff
@@ -591,6 +594,10 @@ describe("Operator contract", (): void => {
 
             expect(formatEther(balanceAfter)).to.equal("179.0")
         })
+
+        it("gets called when withdrawing", async () => {
+            // TODO
```
Comment (jtakalai) [inline packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts:599]: Add test

**Review by jtakalai — CHANGES_REQUESTED** [review]

File: packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts (hunk 3 above)
Comment (samigabor) [reply packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts]: Added

File: packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts [hunk 4]
```diff
@@ -591,6 +594,35 @@ describe("Operator contract", (): void => {
...
+            await expect(operator.stake(sponsorship.address, parseEther("1000")))
+                .to.emit(operator, "Staked").withArgs(sponsorship.address)
+
+            await advanceToTimestamp(timeAtStart + 1000, "Queue for undelegation")
+            await operator.connect(delegator).undelegate(parseEther("100"))
+            // delegator is queued, but no funds were moved yet
+            expect(await token.balanceOf(delegator.address)).to.equal(parseEther("0"))
+            expect(await operator.balanceOf(delegator.address)).to.equal(parseEther("1000"))
+
+            await operator.withdrawEarningsFromSponsorships([sponsorship.address])
```
Comment (jtakalai) [inline packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts:620]: wrap with `.wait()`

File: packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts [hunk 5]
```diff
@@ -591,6 +594,35 @@ describe("Operator contract", (): void => {
...
+            await (await token.connect(sponsor).transferAndCall(sponsorship.address, parseEther("1000"), "0x")).wait() // 1000 available to be earned
+
+            const timeAtStart = await getBlockTimestamp()
+
+            await advanceToTimestamp(timeAtStart, "Stake to sponsorship")
+            await expect(operator.stake(sponsorship.address, parseEther("1000")))
+                .to.emit(operator, "Staked").withArgs(sponsorship.address)
+
+            await advanceToTimestamp(timeAtStart + 1000, "Queue for undelegation")
+            await operator.connect(delegator).undelegate(parseEther("100"))
```
Comment (jtakalai) [inline packages/network-contracts/test/hardhat/OperatorTokenomics/Operator.test.ts:615]: wrap with `.wait()`

**Review by jtakalai — CHANGES_REQUESTED** [review]
👍
